src/main/java/net/snowflake/ingest/streaming/internal/FlushService.java

```diff
@@ -479,17 +480,29 @@ BlobMetadata buildAndUpload(String filePath, List<List<ChannelData>> blobData)
         byte[] compressedChunkData = BlobBuilder.compress(filePath, chunkData);
 
         // Encrypt the compressed chunk data, the encryption key is derived using the key from
-        // server with the full blob path
+        // server with the full blob path.
+        // We need to maintain IV as a block counter for the whole file, even interleaved,
+        // to align with decryption on the Snowflake query path.
+        long iv = curDataSize / Constants.ENCRYPTION_ALGORITHM_BLOCK_SIZE_BYTES;
         byte[] encryptedCompressedChunkData =
-            Cryptor.encrypt(compressedChunkData, firstChannel.getEncryptionKey(), filePath);
+            Cryptor.encrypt(compressedChunkData, firstChannel.getEncryptionKey(), filePath, iv);
+
+        // Encryption adds padding to the ENCRYPTION_ALGORITHM_BLOCK_SIZE_BYTES
+        // to align with decryption on the Snowflake query path starting from this chunk offset.
+        // The padding does not have arrow data and not compressed.
+        // Hence, the actual chunk size is smaller by the padding size.
+        // The compression on the Snowflake query path needs the correct size of the compressed
+        // data.
+        int padding = encryptedCompressedChunkData.length - compressedChunkData.length;
+        int encryptedCompressedChunkDataSize = encryptedCompressedChunkData.length;
+        int chunkLength = encryptedCompressedChunkDataSize - padding;
 
         // SNOW-514965 Do not default to 0 once server side code to send back TMK ID makes it in
         Long encryptionKeyIdToUse =
             firstChannel.getEncryptionKeyId() != null ? firstChannel.getEncryptionKeyId() : 0L;
 
         // Compute the md5 of the chunk data
-        String md5 = BlobBuilder.computeMD5(encryptedCompressedChunkData);
-        int encryptedCompressedChunkDataSize = encryptedCompressedChunkData.length;
+        String md5 = BlobBuilder.computeMD5(encryptedCompressedChunkData, chunkLength);
```

Comment:
When comparing MD5 at server side, instead of comparing MD5 of the entire file/chunk, we will exclude the padding and compute MD5?

Reply:
I guess i'm confusing that why for MD5, we're using `chunkLength`, but for CRC checksum at line 525, we're still using `encryptedCompressedChunkDataSize`

Reply:
True, I did not notice it. I think we should use `chunkLength` for both if they are to be checked/calculated on server side because server side will have only `chunkLength` w/o padding. This is needed for correct decompression.

Reply:
though, crc belongs to the file, right?
then this should be all the bytes I guess
where do we check the crc?

Reply:
> because server side will have only chunkLength w/o padding

Interesting, then how would we figure out the size before decryption? We only use the startOffset?

> where do we check the crc?

hmmm, we don't really use it for now, i think we could leave this as it is. I was just confusing that why we're passing the chunkLength before encryption

Reply:
> Interesting, then how would we figure out the size before decryption? We only use the startOffset?
Decryption picks up the padding automatically because of the block alignment.
e.g. if the block size is 4 and compressed size is 7, the encrypted size will be padded to 8.
if server side asks for file section of 7 bytes, the decryption will assume the same padding, download and decrypt 8 bytes.

decompression has to know the exact size that is 7.

Reply:
thanks, i think i understand why we need the size before padding now and that's been passed as chunk length, but I'm still confused why we're using `compressedChunkLength` to compute the MD5? i assume we just compare MD5 before decompression and decryption?

Reply:
nvm, that saw your comment in the code, it's interesting that we're using the `compressedChunkLength` to compute MD5, instead of `encryptedCompressedChunkDataSize`